libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,313 @@
+}
+
+template<typename It>
+It noReturnValueBegin(It _begin, It _end)
+{
+	assertThrow(_begin != _end, OptimizerException, "expression is empty");
+	AssemblyItem const& topLevelItem = *(_end - 1);
+	if (topLevelItem.returnValues() != 0) return _end;
+	return backExpressionBegin(_begin, _end, true);
+}
+
+vector<vector<AssemblyItem>> parseExpressions(vector<AssemblyItem> const& _items, signed _limit = -1, bool _allowNoReturn = false)
+{
+	auto currentEnd = _items.end();
+	std::vector<vector<AssemblyItem>> expressions{};
+	while (currentEnd != _items.begin() && (_limit == -1 || signed(expressions.size()) < _limit))
+	{
+		auto oldEnd = currentEnd;
+		currentEnd = backExpressionBegin(_items.begin(), oldEnd, _allowNoReturn);
+		expressions.emplace_back(currentEnd, oldEnd);
+		assertThrow(currentEnd >= _items.begin(), OptimizerException, "currentEnd is before begin");
+	}
+
+	return expressions;
+}
+
+vector<vector<AssemblyItem>> parseArguments(vector<AssemblyItem> const& _items)
+{
+	assertThrow(!_items.empty(), OptimizerException, "cannot get arguments from empty _items vector");
+
+	unsigned neededArguments = _items.back().arguments();
+
+	auto expressions = parseExpressions(vector<AssemblyItem>(_items.begin(), _items.end() - 1), neededArguments, false);
+	assertThrow(expressions.size() >= neededArguments, OptimizerException, "more arguments needed than provided");
+
+	expressions = vector<vector<AssemblyItem>>(expressions.begin(), expressions.begin() + neededArguments);
+
+	return expressions;
+}
+
+vector<AssemblyItem> createItems(NewOptimizerPattern const& _pattern)
+{
+	if (_pattern.isBound()) return _pattern.boundItems();
+
+	switch (_pattern.kind())
+	{
+		case NewOptimizerPattern::Kind::Any:
+			assertThrow(false, OptimizerException, "cannot create items from non-bound pattern");
+		case NewOptimizerPattern::Kind::Constant:
+		{
+			if (_pattern.hasConstant())
+				return {_pattern.constant()};
+			else
+				assertThrow(false, OptimizerException, "cannot create items from non-bound constant pattern without constant value");
+		}
+		case NewOptimizerPattern::Kind::Operation:
+			if (_pattern.hasOperationValues())
+			{
+				auto operands = _pattern.operands();
+
+				vector<AssemblyItem> instructions{};
+
+				for (auto it = operands.rbegin(); it != operands.rend(); it++)
+						instructions += createItems(*it);
+
+				instructions.emplace_back(_pattern.instruction());
+
+				return instructions;
+			}
+			else
+				assertThrow(false, OptimizerException, "cannot create items from non-bound operation pattern without operation values");
+		case NewOptimizerPattern::Kind::Unknown:
+			assertThrow(false, OptimizerException, "cannot create items from non-bound unknown pattern");
+		default:
+			assertThrow(false, OptimizerException, "unreachable");
+	}
+}
+
+
+bool dev::solidity::NewOptimizerPattern::matches(std::vector<AssemblyItem> const& _items, bool _unbind)
+{
+	if (_unbind) unbind();
+
+	if (isBound())
+		return boundItems() == _items;
```

Comment:
I am not sure if this is enough. The items have to be identical under any circumstance, and not just consist of the same opcodes.

Reply:
boundItems consists of all of the items, not just opcodes. If AssemblyItem's operator== compares data (which it does), then this should check if items are exactly identical.

Reply:
Sure, but what about e.g. `dup3` or `sload`? Both depend on the context and just comparing them is not enough.

Reply:
`dup3` would only exist with arguments, so that would be fine.  `sload` is more tricky, but since it has side effects, it would never have rules applied against it in this module. And, it would only match if both the location and value were the same.